### PR TITLE
Develop logging

### LINKED ISSUE #27: 🚀 Feature: Enhanced Logging
*As a user, I'd like to have enhanced logging capabilities including slack notifications....*

cc: @lmz to fill out more

### PR DESCRIPTION
**Pull request by inrokevin**
## Adds enhanced logging capabilities
Closes #27 

Some notes on using the logger (cross referencing the config below).

 - Adds logging with levels for display, file and fallback log everything on error plus "override" logging level filter by component name and iteration, and notify slack component (untested at this time)
 - Log messages can be shown in the console / notebook (based on the logging.display_level)
 - There are two main log files used: 1. a "run" log which is meant to provide model overview (logging.run_file_path, and logging.run_file_level) and 2. a more detailed log (logging.log_file_path, logging.log_file_level) 
 - There is also the logging.log_on_error_file_path which will output all log messages recorded.
 - Note that some logging can be conditional to only run if the log level is filtered in, e.g. if it takes a long time to generate the report. There is an example of this in the highway assignment which generates a report of the matrix results statistics only if DEBUG is filtered in for at least one of the log_levels
 - The logging.iter_component_level can be used to locally override the logging level filter for debug purposes. This is specified as one or more tuples of (iteration, component_name, log_level), e.g. [ [2, "highway", "TRACE"] ] to record all messages during the highway component run at iteration 2.

For reference here are the log levels as I have defined them. 
- TRACE: highly detailed information which would rarely be of interest except for detailed debugging by a developer.
- DEBUG: diagnostic information which would generally be useful to a developer debugging the model code; this may also be useful to a model operator in some cases.
- DETAIL: more detail than would normally be of interest, but might be useful to a model operator debugging a model run / data or understanding model results.
- INFO: messages which would normally be worth recording about the model operation.
- STATUS: top-level, model is running type messages. There should be relatively few of these, generally one per component, or one per time period if the procedure is long.
- WARN: warning messages where there is a possibility of a problem.
- ERROR: problem causing operation to halt which is normal (or not unexpected) in scope, e.g. file does not exist. Includes general Python exceptions.
- FATAL: severe problem requiring operation to stop immediately.

(In practice I suspect there may not be a distinction between ERROR and FATAL in tm2py context.)

Messages can be recorded using logger.log(<message> level="INFO") or equivalently logger.info(<message>). You can include time=True, or indent=True to add a timestamp or indent the messages in the text file, or shortcut the former with logger.log_time(<message>, level=<level>)

You can also use the context with logger.log_start_end(<message>) to group the log messages together. And there is an equivalent decorator which can be used on methods:

    @LogStartEnd("Highway assignment and skims", level="STATUS")
    def run(self):

## Testing

 - [x] Code for this PR is covered in tests 
 - [x] Code passes all existing tests
 - [x] Doc for logging added, including examples
 - [x] Code linted

### CODE REVIEW

File: tests/test_logging.py
Comment (e-lo) [inline tests/test_logging.py:61]: why is this allowed to be passing?

File: tests/test_logging.py
Comment (e-lo) [inline tests/test_logging.py:64]: I"m not exactly sure what this is testing? Might be helpful to add comments.

File: tests/test_logging.py
Comment (e-lo) [inline tests/test_logging.py:38]: avoid try/excepts in tests generally - unless you are testing that an exception if caught

File: tests/test_logging.py
Comment (e-lo) [inline tests/test_logging.py:11]: this doesn't have to be in tempfile context

File: tests/test_logging.py
Comment (e-lo) [inline tests/test_logging.py:23]: i don't htink this has to be in tempfile context

File: tests/test_logging.py
Comment (e-lo) [inline tests/test_logging.py:39]: this is two-contexts deep...that gets dicy fast.

File: tm2py/controller.py
Comment (e-lo) [inline tm2py/controller.py:128]: I"m not sure I love having to use the context?  Is there a reason for it?  cc @lmz who may also have thoughts here.

**Review by e-lo — CHANGES_REQUESTED** [review]
Things generally good/complete. My main holdbacks are:
1. Having to use contexts (cc: @LMZ for her thoughts here - if she's ok, i'm good)
2. Not sure what the tests are testing.  See detailed comments there.

File: tests/test_logging.py
Comment (inrokevin) [reply tests/test_logging.py:61]: In this case testing that the behaviour on error is correct, that an error message is recorded in the log and the log_on_error file is generated. Added some comments for the record.

File: tests/test_logging.py
Comment (inrokevin) [reply tests/test_logging.py:38]: See above, testing logging in case of error.

File: tests/test_logging.py
Comment (inrokevin) [reply tests/test_logging.py:64]: Added some comments, checking the content of the log files, hopefully clear for the future.

File: tests/test_logging.py
Comment (inrokevin) [reply tests/test_logging.py:11]: Moved

File: tm2py/controller.py
Comment (e-lo) [reply tm2py/controller.py:128]: I'll defer to @lmz  - generally I don't love having to indent everything that I want to log

Comment (e-lo) [reply tm2py/controller.py:128]: (granted I realize that you can do it on the calling function)

Comment (lmz) [reply tm2py/controller.py:128]: Sorry I'm behind on this... 
I don't like having to indent everything that I want to log either.

**Comment by lmz** [comment]
> @lmz - given that Kevin is out for now can we go ahead and merge this?

Yes, I don't want to hold it up.
I need to get things running (!) so I can see how this all fits together but if we make changes, we can do so after the merge.

File: tm2py/controller.py
Comment (e-lo) [reply tm2py/controller.py:128]: I'll create a separate issue

**Pull request merged**
